### PR TITLE
 TextPosition.py: Overload < and > operators

### PR DESCRIPTION
**Pull request by aptrishu**
It overloads __le__ and __ge__ operators
in TextPosition.py to explicitly check
for None values.

Fixes https://github.com/coala/coala/issues/3094

### Checklist

- [x] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [x] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [x] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [x] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [x] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [x] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

**After you submit your pull request, DO NOT click the 'Update Branch' button**

### CODE REVIEW

**Comment by aptrishu** [comment]
It also makes some changes in ProcessingTest.py and ResultTest.py because of new overlaps() method.

**Comment by aptrishu** [comment]
CC @sils @Makman2

**Comment by aptrishu** [comment]
Earlier we were not checking for file equality because SourcePosition objects were ordered according to file but now since I'm doing no direct comparison so we'll have to check for file equality explicitly.

File: coalib/results/SourceRange.py [hunk 1]
```diff
@@ -158,6 +158,10 @@ def __str__(self):
 
         return format_str.format(self)
 
+    def overlaps(self, other):
+        return (super().overlaps(other) and
+                self.start.file == other.start.file)
+
```
Comment (sils) [inline coalib/results/SourceRange.py]: wooow this was missing? Can we have an own commit for this?

File: coalib/results/TextRange.py [hunk 2]
```diff
@@ -89,7 +89,31 @@ def end(self):
...
+                if self.end.column is None or other.start.column is None:
+                    second = True
+                else:
+                    second = self.end.column >= other.start.column
+            else:
+                second = self.end.line > other.start.line
+        else:
+            second = True
+
+        return first and second
```
Comment (sils) [inline coalib/results/TextRange.py]: are you sure this has to be so complicated?

File: coalib/results/TextRange.py [hunk 3]
```diff
@@ -89,7 +89,31 @@ def end(self):
         return self._end
 
     def overlaps(self, other):
-        return self.start <= other.end and self.end >= other.start
```
Comment (sils) [inline coalib/results/TextRange.py]:
```python
if self.start.line is None or other.start.line is None:  # End cannot be None if start isn't
    return self.file == other.file

if self.start.column is None or other.start.column is None:
    # compare lines

return self.start <= other.end and self.end >= other.start
```

should do it shorter.

File: coalib/results/TextRange.py (hunk 2 above)
Comment (aptrishu) [inline coalib/results/TextRange.py]: I was also thinking about it. I'll update it (if possible)

File: coalib/results/SourceRange.py (hunk 1 above)
Comment (aptrishu) [inline coalib/results/SourceRange.py]: @fneu had removed this check, as SourcePositions are already ordered by file.

Comment (sils) [inline coalib/results/SourceRange.py]: no additional test? Esp if this fixes something regression test would be good

File: coalib/results/TextRange.py [hunk 4]
```diff
@@ -89,7 +89,31 @@ def end(self):
...
+        first = False
+        second = False
```
Comment (sils) [inline coalib/results/TextRange.py]: can we have better names for first and second? What do they stand for really?

File: coalib/results/SourceRange.py (hunk 1 above)
Comment (aptrishu) [inline coalib/results/SourceRange.py]: Ok, I'll add some tests. :+1:

File: coalib/results/TextRange.py (hunk 4 above)
Comment (aptrishu) [inline coalib/results/TextRange.py]: first stands for self.start <= other.end
second stands for self.end >= other.start

Comment (sils) [inline coalib/results/TextRange.py]: I think that should go into a comment at least

File: tests/results/SourceRangeTest.py [hunk 5]
```diff
@@ -103,6 +103,32 @@ def test_contains(self):
         b = SourceRange.from_values('test_file', 1, 1, 50, 20)
         self.assertNotIn(a, b)
 
+    def test_overlaps(self):
```
Comment (Udayan12167) [inline tests/results/SourceRangeTest.py]: These are tests for sourcerange shouldn't they be in the other commit?

File: coalib/results/TextRange.py (hunk 3 above)
Comment (Udayan12167) [inline coalib/results/TextRange.py]: 👍

File: tests/results/ResultTest.py [hunk 6]
```diff
@@ -139,6 +139,20 @@ def test_overlaps(self):
         self.assertTrue(uut.overlaps([overlapping_range]))
         self.assertFalse(uut.overlaps(nonoverlapping_range))
 
+        overlapping_range = SourceRange.from_values('file1', 1, None, 1, None)
```
Comment (Udayan12167) [inline tests/results/ResultTest.py]: Again I feel this is related to sourcerange commit

File: tests/processes/ProcessingTest.py [hunk 7]
```diff
@@ -219,8 +219,7 @@ def test_process_queues(self):
             self.log_printer,
             self.console_printer)
 
-        self.assertEqual(self.queue.get(timeout=0), ([first_local,
-                                                      second_local,
+        self.assertEqual(self.queue.get(timeout=0), ([second_local,
```
Comment (Udayan12167) [inline tests/processes/ProcessingTest.py]: Unrelated change?

**Comment by aptrishu** [comment]
@Udayan12167 yeah, I have also included sourcerange and result tests in the same commit because they test the same function `overlaps()`, so they were needed to be tested. Shall I put everyone of them in extra commits ? we'd have 3 extra commits just for tests.

File: tests/processes/ProcessingTest.py (hunk 7 above)
Comment (aptrishu) [inline tests/processes/ProcessingTest.py]: No, the new overlaps method covers `first_local` in ignore range.

File: coalib/results/TextRange.py (hunk 3 above)
Comment (aptrishu) [inline coalib/results/TextRange.py]: As discussed earlier on gitter, we should go with the long method I suggested because we need to handle None values properly..

**Comment by aptrishu** [comment]
moved the sourceRange tests changes to sourcerange commit.. :+1: @Udayan12167

**Comment by Makman2** [comment]
I think you should rebase, want to see all lights go green here :3

**Comment by aptrishu** [comment]
Rebased.. @Makman2

File: coalib/results/TextRange.py [hunk 8]
```diff
@@ -89,7 +89,35 @@ def end(self):
         return self._end
 
     def overlaps(self, other):
-        return self.start <= other.end and self.end >= other.start
+        first = False  # check for left boundary overlap
+        # self.start <= other.end
+
+        second = False  # check for right boundary overlap
+        # self.end >= other.start
```
Comment (Makman2) [inline coalib/results/TextRange.py]: why is it code commented out?

File: coalib/results/TextRange.py [hunk 9]
```diff
@@ -89,7 +89,35 @@ def end(self):
...
+
+        if None not in (self.start.line, other.end.line):
```
Comment (Makman2) [inline coalib/results/TextRange.py]: I would refactor to `if self.start.line is not None and other.end.line is not None` -> faster and imo slightly better to understand.

But up to you :+1:

File: coalib/results/TextRange.py [hunk 10]
```diff
@@ -89,7 +89,35 @@ def end(self):
...
+            if self.start.line == other.end.line:
+                if None in (self.start.column, other.end.column):
+                    first = True  # if column = None, consider whole line
+                else:
+                    first = self.start.column <= other.end.column
```
Comment (Makman2) [inline coalib/results/TextRange.py]: -->
```
first = True if None in (...) else self.start.column <= other.end.column
```

File: coalib/results/TextRange.py [hunk 11]
```diff
@@ -89,7 +89,35 @@ def end(self):
...
+        if None not in (self.end.line, other.start.line):
+            if self.end.line == other.start.line:
+                if None in (self.end.column, other.start.column):
+                    second = True
+                else:
+                    second = self.end.column >= other.start.column
+            else:
+                second = self.end.line > other.start.line
+        else:
+            second = True
```
Comment (Makman2) [inline coalib/results/TextRange.py]: maybe you can make out of these if blocks a function, like `is_behind(first_position, second_position)`, and then do:
```
return is_behind(...) and is_behind(...)
```

File: coalib/results/TextRange.py (hunk 2 above)
Comment (Makman2) [inline coalib/results/TextRange.py]: actually this looks right... don't know myself how to simplify, it correctly accounts for the cases with `None`.... wouldn't have done it differently^^

File: coalib/results/TextRange.py (hunk 8 above)
Comment (aptrishu) [inline coalib/results/TextRange.py]: Actually, @sils asked me to comment 'what first and second represents'.

File: coalib/results/TextRange.py (hunk 10 above)
Comment (aptrishu) [inline coalib/results/TextRange.py]: It is split up so that I could comment on the if part. So, its just for the sake of clarity.

File: coalib/results/TextRange.py (hunk 11 above)
Comment (aptrishu) [inline coalib/results/TextRange.py]: would It not make the `overlaps` method complicated ?
I have put comments to clarify `first` and `second`.

File: coalib/results/TextRange.py (hunk 8 above)
Comment (sils) [inline coalib/results/TextRange.py]: @aptrishu usually the best thing in those cases is better variable names

**Comment by aptrishu** [comment]
@Makman2 @sils Does it look fine now ?

File: coalib/results/TextRange.py [hunk 12]
```diff
@@ -89,7 +89,19 @@ def end(self):
...
-        return self.start <= other.end and self.end >= other.start
+        return self.is_behind(self.start, other.end) and \
+            self.is_behind(other.start, self.end)
+
+    @staticmethod
+    def is_behind(first_position, second_position):
+        if first_position.line is None or second_position.line is None:
+            return True
+        if first_position.line == second_position.line:
+            return True if None in \
```
Comment (Mixih) [inline coalib/results/TextRange.py]: Stylistic thing, but I think we prefer parenthesis syntax over the escaped char one `\` (like return(...))

File: coalib/results/SourceRange.py [hunk 13]
```diff
@@ -158,6 +158,10 @@ def __str__(self):
 
         return format_str.format(self)
 
+    def overlaps(self, other):
+        return (super().overlaps(other) and
+                self.start.file == other.start.file)
```
Comment (Makman2) [inline coalib/results/SourceRange.py]: I would put the `self.start.file == other.start.file` before `super().overlaps(other)`, the file-equality check is not so expensive than the base-overlaps check :)

File: coalib/results/TextRange.py [hunk 14]
```diff
@@ -89,7 +89,19 @@ def end(self):
         return self._end
 
     def overlaps(self, other):
-        return self.start <= other.end and self.end >= other.start
+        return self.is_behind(self.start, other.end) and \
+            self.is_behind(other.start, self.end)
```
Comment (Makman2) [inline coalib/results/TextRange.py]: please use braces instead of `\` :)

File: coalib/results/TextRange.py [hunk 15]
```diff
@@ -89,7 +89,19 @@ def end(self):
...
+
+    @staticmethod
+    def is_behind(first_position, second_position):
+        if first_position.line is None or second_position.line is None:
+            return True
+        if first_position.line == second_position.line:
+            return True if None in \
+                (first_position.column, second_position.column) \
+                else first_position.column <= second_position.column
```
Comment (Makman2) [inline coalib/results/TextRange.py]: up to you:
if if's or for's do wrap onto multiple lines, I wrap lines according to the different meanings of the stmt, e.g.
```
return-value 1
if condition else
return-value 2
```
-->
```python
True
if None in (first_position.column, second_position.column) else
first_position.column <= second_position.column
```
imo this is better to read, but like said, up to you :)

**Comment by aptrishu** [comment]
review again.. -_- @Makman2

File: tests/results/TextRangeTest.py [hunk 16]
```diff
@@ -87,6 +87,26 @@ def test_overlap(self):
...
+
+        uut1 = TextRange.from_values(None, None, 1, None)
+        uut2 = TextRange.from_values(2, 1, 3, None)
+        self.assertFalse(uut1.overlaps(uut2))
+        self.assertFalse(uut2.overlaps(uut1))
+
+        uut1 = TextRange.from_values(None, None, None, None)
+        uut2 = TextRange.from_values(None, None, None)
+        self.assertTrue(uut1.overlaps(uut2))
+        self.assertTrue(uut2.overlaps(uut1))
```
Comment (Makman2) [inline tests/results/TextRangeTest.py]: could you make a testcase for different lines but `None` columns?

Comment (aptrishu) [inline tests/results/TextRangeTest.py]: uut1 = TextRange.from_values(1, None, 1, None)
uut2 = TextRange.from_values(2, None, 2, None)
like this ?

Comment (Makman2) [inline tests/results/TextRangeTest.py]: yeah for example :)

File: coalib/results/TextRange.py [hunk 17]
```diff
@@ -89,7 +89,20 @@ def end(self):
...
+        return (self.is_behind(self.start, other.end) and
+                self.is_behind(other.start, self.end))
+
+    @staticmethod
+    def is_behind(first_position, second_position):
+        if first_position.line is None or second_position.line is None:
+            return True
+        if first_position.line == second_position.line:
+            return (True
+                    if None in (first_position.column, second_position.column)
```
Comment (Makman2) [inline coalib/results/TextRange.py]: you should expand this one too to `if first_position.column is None or ...`, I tested this with `timeit`:

```python
>>> from timeit import *
>>> def in_approach(x, y):
...  if None in (x, y):
...   return 3
...  else:
...   return 4
... 
>>> def direct_approach(x, y):
...  if x is None or y is None:
...   return 3
...  else:
...   return 4
... 
>>> data = []
>>> for i in range(100):
...  if i % 3 == 0:
...   data.append((None, i))
...  elif i % 7 == 0:
...   data.append((i, None))
...  else:
...   data.append((i, i))
... 
>>> def IN():
...  for x, y in data:
...   in_approach(x, y)
... 
>>> def DIRECT():
...  for x, y in data:
...   direct_approach(x, y)
... 
>>> T_IN = Timer(IN)
>>> T_DIRECT = Timer(DIRECT)
>>> count = 100000
>>> r = (T_IN.timeit(count), T_DIRECT.timeit(count))
>>> r
(1.7751383920000308, 1.2874540749999142)
>>> r[1] / r[0]
0.698429337908
```
You see it's about 30% faster, this is quite much^^

File: coalib/results/TextRange.py [hunk 18]
```diff
@@ -89,7 +89,20 @@ def end(self):
...
+                    else
+                    first_position.column <= second_position.column)
```
Comment (Makman2) [inline coalib/results/TextRange.py]: maybe for the sake of readability, use normal if's again, not sure any more... up to you :)

File: coalib/results/TextRange.py [hunk 19]
```diff
@@ -89,7 +89,20 @@ def end(self):
         return self._end
 
     def overlaps(self, other):
-        return self.start <= other.end and self.end >= other.start
+        return (self.is_behind(self.start, other.end) and
+                self.is_behind(other.start, self.end))
+
+    @staticmethod
+    def is_behind(first_position, second_position):
```
Comment (Makman2) [inline coalib/results/TextRange.py]: what about actually overriding the `<` and `>` operators for `TextPosition` with your code? Then you can just do inside `overlaps`, then you can keep the old method?

**Comment by Makman2** [comment]
ack 58f9160

**Comment by Makman2** [comment]
unack 5d2775f

**Comment by aptrishu** [comment]
@Makman2 All changes done.

File: coalib/results/TextPosition.py [hunk 20]
```diff
@@ -31,3 +31,27 @@ def line(self):
     @property
     def column(self):
         return self._column
+
+    def __le__(self, other):
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: not `__lt__` and `__gt__`?
As your last comparison does
```
return self.line < other.line
```

File: tests/results/TextRangeTest.py [hunk 21]
```diff
@@ -87,6 +87,31 @@ def test_overlap(self):
...
+
```
Comment (Makman2) [inline tests/results/TextRangeTest.py]: Could you make also a test inside `TestPosition` as well?

**Comment by Makman2** [comment]
ack 901c5b0

File: coalib/results/TextPosition.py (hunk 20 above)
Comment (aptrishu) [inline coalib/results/TextPosition.py]: no, I wrote `__le__` and `__ge__` to handle
`self.start <= other.end and self.end >= other.start` in TextRange.

File: tests/results/TextRangeTest.py (hunk 21 above)
Comment (aptrishu) [inline tests/results/TextRangeTest.py]: sure.. :+1:

**Comment by aptrishu** [comment]
@Makman2 updated.. :+1:  and rebased too.. :P

File: coalib/results/TextPosition.py [hunk 22]
```diff
@@ -31,3 +31,27 @@ def line(self):
...
+        if self.line == other.line:
+            return (True
+                    if self.column is None or  # if column is None,
+                    other.column is None  # consider whole line
+                    else
+                    self.column <= other.column)
+        else:
+            return self.line < other.line
+
+    def __ge__(self, other):
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: you could actually document their behaviours, not so easy when `None` comes into play :)

Comment (aptrishu) [inline coalib/results/TextPosition.py]: I have put comments there in __le__.

File: tests/results/TextPositionTest.py [hunk 23]
```diff
@@ -30,3 +30,29 @@ def test_properties(self):
...
+
+        uut1 = TextPosition(None, None)
+        uut2 = TextPosition(4, 5)
+        self.assertTrue(uut1 >= uut2)
+        self.assertTrue(uut2 >= uut1)
+
+        uut1 = TextPosition(4, 8)
+        uut2 = TextPosition(4, 8)
+        self.assertTrue(uut1 >= uut2)
+        self.assertTrue(uut2 >= uut1)
```
Comment (Makman2) [inline tests/results/TextPositionTest.py]: where's `<=`? :O

Comment (aptrishu) [inline tests/results/TextPositionTest.py]: ugh.. damn.. How did I miss that.. :( updating..

File: coalib/results/TextPosition.py [hunk 24]
```diff
@@ -31,3 +31,27 @@ def line(self):
...
     def column(self):
         return self._column
+
+    def __le__(self, other):
+        if self.line is None or other.line is None:
+            return True  # if line is None, consider whole file
+        if self.line == other.line:
+            return (True
+                    if self.column is None or  # if column is None,
+                    other.column is None  # consider whole line
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: when you document properly, I think you don't need those comments then :)
(especially as you don't use them below 😄 )

Comment (aptrishu) [inline coalib/results/TextPosition.py]: Yeah, I'm gonna document them..

**Comment by aptrishu** [comment]
@Makman2 It's ready. :)

File: coalib/results/TextPosition.py [hunk 25]
```diff
@@ -31,3 +31,43 @@ def line(self):
     @property
     def column(self):
         return self._column
+
+    def __le__(self, other):
+        """
+        Performs comparison between two TextPositions, logically equivalent
+        to (self <= other). If the column in a TextPosition is None, consider
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: why braces?

File: coalib/results/TextPosition.py [hunk 26]
```diff
@@ -31,3 +31,43 @@ def line(self):
...
+        whole line. If the line in a TextPosition is None, consider whole file.
+
+        :param other: TextPosition to compare with.
+        :return:      Boolean value logically equivalent to (self <= other).
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: Enclose `TextPosition` in double backticks :)

Comment (Makman2) [inline coalib/results/TextPosition.py]: Also I think you shuld rephrase the `:return:` description to something like 
```
Whether this ``TextPosition`` is behind the other one or the same.
```
or similar :)

**Comment by aptrishu** [comment]
Updated. @Makman2

File: coalib/results/TextPosition.py [hunk 27]
```diff
@@ -31,3 +31,46 @@ def line(self):
     @property
     def column(self):
         return self._column
+
+    def __le__(self, other):
+        """
+        Performs comparison between two TextPositions, logically equivalent
+        to self <= other. If the column in a TextPosition is None, consider
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: "logically equivalent to self <= other": hm yeah, kinda obvious, that's what `__le__` is for ;D

Maybe just write "Test whether ``self`` is behind the ``other`` Textposition" or so :)

Comment (Makman2) [inline coalib/results/TextPosition.py]: Also enclose `None`s in double backticks ;)

**Comment by aptrishu** [comment]
Updated. @Makman2 
Issue filed https://github.com/coala/coala/issues/3809.

File: coalib/results/TextPosition.py [hunk 28]
```diff
@@ -31,3 +31,48 @@ def line(self):
     @property
     def column(self):
         return self._column
+
+    def __le__(self, other):
+        """
+        Test whether self is behind the other TextPosition.
+        If the column in a TextPosition is ``None``, consider
+        whole line. If the line in a TextPosition is ``None``,
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: TextPosition in double backticks please ;) (counts for all occurrences)

File: coalib/results/TextPosition.py [hunk 29]
```diff
@@ -31,3 +31,48 @@ def line(self):
     @property
     def column(self):
         return self._column
+
+    def __le__(self, other):
+        """
+        Test whether self is behind the other TextPosition.
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: I think you can add a newline here :+1:

Comment (Makman2) [inline coalib/results/TextPosition.py]: `self` could be also enclosed in double backticks :) (counts for all occurrences)

File: coalib/results/TextPosition.py [hunk 30]
```diff
@@ -31,3 +31,48 @@ def line(self):
...
+                    if self.column is None or
+                    other.column is None
+                    else
+                    self.column <= other.column)
+        else:
+            return self.line < other.line
+
+    def __ge__(self, other):
+        """
+        Test whether self is ahead of the other TextPosition.
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: Also here I would insert a newline.

**Comment by aptrishu** [comment]
Updated. @Makman2 
Sorry If this PR is taking too long. seems I'm real bad at documenting things.. :D

**Comment by Makman2** [comment]
No hurry :)

File: coalib/results/TextPosition.py [hunk 31]
```diff
@@ -9,8 +9,8 @@ class TextPosition:
     @enforce_signature
     def __init__(self, line: (int, None)=None, column: (int, None)=None):
         """
-        Creates a new TextPosition object that represents the position inside
-        a string with line/column numbers.
+        Creates a new ``TextPosition`` object that represents the position
+        inside a string with line/column numbers.
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: separate commit :)

File: coalib/results/TextPosition.py [hunk 32]
```diff
@@ -31,3 +31,50 @@ def line(self):
     @property
     def column(self):
         return self._column
+
+    def __le__(self, other):
+        """
+        Test whether ``self`` is behind the other ``TextPosition``.
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: ah I would add "... is behind **or equals** the other ...". Same for `__ge__` below :3

File: coalib/results/TextPosition.py (hunk 31 above)
Comment (aptrishu) [inline coalib/results/TextPosition.py]: removed this change.. I'll update this in other PR (the one implementing _lt_, __gt__)

**Comment by aptrishu** [comment]
Updated. @Makman2

File: coalib/results/TextPosition.py [hunk 33]
```diff
@@ -9,8 +9,8 @@ class TextPosition:
     @enforce_signature
     def __init__(self, line: (int, None)=None, column: (int, None)=None):
         """
-        Creates a new TextPosition object that represents the position inside
-        a string with line/column numbers.
+        Creates a new TextPosition object that represents the position
+        inside a string with line/column numbers.
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: unrelated change :)

**Comment by aptrishu** [comment]
updated. @Makman2

File: coalib/results/TextPosition.py [hunk 34]
```diff
@@ -31,3 +31,52 @@ def line(self):
...
+        other ``TextPosition``.
+
+        If the column in a ``TextPosition`` is ``None``, consider
+        whole line. If the line in a ``TextPosition`` is ``None``,
+        consider whole file.
+
+        :param other: ``TextPosition`` to compare with.
+        :return:      Whether this ``TextPosition`` is ahead of the other
+                      one or the same.
+
```
Comment (Makman2) [inline coalib/results/TextPosition.py]: redundant newline :3

**Comment by aptrishu** [comment]
Updated. @Makman2

**Comment by Makman2** [comment]
ack f6f00ea 8473890

**Comment by Makman2** [comment]
rebase and lgtm ;)

**Comment by aptrishu** [comment]
ack 00d383d
ack c395d6e

**Comment by aptrishu** [comment]
ack 00d383d4c67c7ce981f65af0196349c7639c21f1

**Comment by aptrishu** [comment]
ack c395d6e6849855e08e824810e572e419edbf2011

**Comment by aptrishu** [comment]
@rultor merge
